# Task: Linear vs Polynomial Regression

## Objective
Build and compare Linear Regression and Polynomial Regression models then select the best performing model and interpret its learned equation.


## Dataset
Use the **Boston Housing dataset**.
    
    from sklearn.datasets import load_boston

In [80]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.datasets import fetch_california_housing

In [81]:
housing = fetch_california_housing()
X = pd.DataFrame(housing.data, columns=housing.feature_names)
y = pd.Series(housing.target, name='MedHouseVal')

print('Dataset shape:', X.shape)
print('Target shape :', y.shape)
print('\nFeatures:', list(X.columns))
X.head()

Dataset shape: (20640, 8)
Target shape : (20640,)

Features: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25


### 1. Data Preparation

In [82]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### 2. Linear Regression Model
- Train a Linear Regression model
- Make predictions on the test set
- Compute evaluation metrics:
  - MAE
  - MSE
  - RMSE

In [83]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [84]:
model.fit(X_train,y_train)

LinearRegression()

In [85]:
model.coef_

array([ 4.48674910e-01,  9.72425752e-03, -1.23323343e-01,  7.83144907e-01,
       -2.02962058e-06, -3.52631849e-03, -4.19792487e-01, -4.33708065e-01])

In [86]:
pd.Series(model.coef_, index=X.columns)

MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64

In [87]:
model.intercept_

np.float64(-37.02327770606416)

In [88]:
test_predictions = model.predict(X_test)

In [89]:
from sklearn.metrics import mean_absolute_error,mean_squared_error
MAE = mean_absolute_error(y_test,test_predictions)
MSE = mean_squared_error(y_test,test_predictions)
RMSE = np.sqrt(MSE)

In [90]:
print(MAE, MSE, RMSE)

0.533200130495656 0.555891598695244 0.7455813830127761


### 3. Polynomial Regression Model
- Train a Polynomial Regression model
- Make predictions on the test set
- Compute evaluation metrics:
  - MAE
  - MSE
  - RMSE


In [91]:
from sklearn.preprocessing import PolynomialFeatures
polynomial_converter = PolynomialFeatures(degree=2)

poly_features = polynomial_converter.fit_transform(X)

In [92]:
X_train, X_test, y_train, y_test = train_test_split(poly_features, y, test_size=0.2, random_state=42)

In [93]:
modelpoly = LinearRegression()

In [94]:
modelpoly.fit(X_train,y_train)


LinearRegression()

In [95]:
test_predictions = modelpoly.predict(X_test)

In [96]:
MAEP = mean_absolute_error(y_test,test_predictions)
MSEP = mean_squared_error(y_test,test_predictions)
RMSEP = np.sqrt(MSEP)


print(f"MAE: {MAEP}")
print(f"MSE: {MSEP}")
print(f"RMSE: {RMSEP}")

MAE: 0.46700093347739485
MSE: 0.46430152634628236
RMSE: 0.6813967466507911


### 4. Model Comparison
- Compare Linear vs Polynomial Regression results
- Identify which model performs better on the test data and justify your decision


In [98]:
# Summary comparison table
comparison = pd.DataFrame({
    'Model' : ['Linear Regression', 'Polynomial Regression'],
    'MAE'   : [round(MAE,4),   round(MAEP,4)],
    'MSE'   : [round(MSE,4),   round(MSEP,4)],
    'RMSE'  : [round(RMSE,4),  round(RMSEP,4)],
})
print(comparison.to_string(index=False))

                Model    MAE    MSE   RMSE
    Linear Regression 0.5332 0.5559 0.7456
Polynomial Regression 0.4670 0.4643 0.6814


Best Model : Polynomial Regression
  RMSE  Linear Regression      : 0.7456
  RMSE  Polynomial Regression  : 0.6814
  Improvement                  : 8.61%

  • Polynomial Regression adds squared terms and cross-feature interactions (e.g. MedInc × HouseAge), letting the model capture non-linear relationships in the data.

  • Lower MAE, MSE, and RMSE on the test set confirm that it generalises better than plain Linear Regression.

  • Degree = 2 improves fit without the severe overfitting risk of higher degrees.